### PR TITLE
Add utils to build web servlets

### PR DESCRIPTION
**Pull request by babolivier**
Lifted almost verbatim from https://github.com/matrix-org/sydent/blob/main/sydent/http/servlets/__init__.py

The reason behind this PR is that I'm looking at rewriting https://github.com/matrix-org/matrix-content-scanner in Python and I realised a lot of code to help creating and managing web servlets lives in Sydent, and I'd rather move it in here than copy it over to a new project.

I haven't added tests, mostly because I'm not sure how to test this kind of things.

### CODE REVIEW

**Comment by clokep** [comment]
Would it make more sense to use the `JsonResource` (which has a wholeeeeee bunch of other stuff it depends on, unfortunately) from Synapse instead? I'm much more confident that code works properly than what's in Sydent.

**Comment by babolivier** [comment]
> Would it make more sense to use the `JsonResource` (which has a wholeeeeee bunch of other stuff it depends on, unfortunately) from Synapse instead? I'm much more confident that code works properly than what's in Sydent.

Not sure, I feel like Synapse's `JsonResource` might be a bit overengineered for just a use case of "I want to make a servlet that renders JSON".

**Comment by clokep** [comment]
Yeah, that's fair. I guess I like the way `_AsyncResource` works more than using the decorators? Not a big deal though.

File: matrix_common/servlet.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                    "error": "Internal Server Error",
+                }
+            )
+
+    return inner
+
+
+def json_servlet_async(fn: Callable[..., JsonDict]) -> Callable[..., int]:
+    async def render(
+        fn: Callable[..., Any], self: Any, request: Request, **kwargs: Any
```
Comment (squahtx) [inline matrix_common/servlet.py:82]: We await and json encode the return from `fn` below, so we probably want it to return `Awaitable[JsonDict]`?

```suggestion
        fn: Callable[..., Awaitable[JsonDict]], self: Any, request: Request, **kwargs: Any
```

File: matrix_common/servlet.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        """
+        Runs an asynchronous web handler function with the given request and parameters,
+        then converts its result into JSON bytes and writes it to the request. If an error
+        happens, also sets the HTTP response code.
+
+        Args:
+            fn: The handler to run.
+            self: The current object.
+            request: The request to process.
+            args: The arguments to pass to the function.
```
Comment (squahtx) [inline matrix_common/servlet.py:93]: Is the `*args` argument missing?

File: matrix_common/servlet.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+from twisted.web.server import Request
+
+from matrix_common.json import JsonDict
+
+logger = logging.getLogger(__name__)
+
+
+class MatrixRestError(Exception):
+    """
+    Handled by the jsonwrap wrapper. Any servlets that don't use this
```
Comment (squahtx) [inline matrix_common/servlet.py:32]: Does `jsonwrap` refer to the decorators below?

File: matrix_common/servlet.py (hunk 1 above)
Comment (DMRobertson) [reply matrix_common/servlet.py:82]: C.f. what I did for Sydent here: https://github.com/matrix-org/sydent/blob/143a22520b974ac63987de8dc048891071f299e4/sydent/http/servlets/__init__.py#L177-L181

File: matrix_common/servlet.py (hunk 3 above)
Comment (babolivier) [reply matrix_common/servlet.py:32]: Err, yes, I forgot to rename it here.

**Comment by babolivier** [comment]
After having progressed a bit more into the content scanner rewrite, I'm not sure this is how I want it this to look like. I'll revisit once I'm back from holidays.

**Comment by babolivier** [comment]
Closing this as we'll probably want to go another way.

**Pull request closed without merging**